### ISSUE TITLE
New-OVLdapServer with the TrustLeafCertificate switch set fails

### ISSUE THREAD

**Issue by erikgraa** [HewlettPackard/POSH-HPOneView#518]
### Expected Behavior

New-OVLdapServer -Hostname 'domaincontroller.fqdn' -TrustLeafCertificate

**NOTE**: Adding the directory and all domain controllers works perfectly through the UI.

### Actual Behavior

New-OVLdapServer -Hostname 'domaincontroller.fqdn' -TrustLeafCertificate

Fails with:

![image](https://user-images.githubusercontent.com/20067955/98709934-4f1e3600-2383-11eb-99f2-d5dad5714910.png)

How can I know which algorithms are being attempted/negotiated?

### Steps to reproduce

Domain controller certificate has SHA384RSA algorithm and RSA 2048 bits public key.

### Version Information
HPE OneView PowerShell Library Version (Get-HPOVVersion or $PSLibraryVersion): 5.50.2591.2980
HPE OneView Appliance Version (Get-HPOVVersion -ApplianceVer): 5.50.00.426657.00
Output from $PSVersionTable on your Windows Host: 5.1.17763.1432

**Comment by ChrisLynchHPE** [comment]
There is no special negotiation here.  This is dependent upon Microsoft's `System.Net.Sockets.TcpClient` class to download the certificate to then pass as a property value for the REST API.

Two things:

1. Is there a reason why you are not adding your CA root chain as appliance trusted certificates?  By adding the Domain Controllers leaf cert, you then have to update the certificate from the DC when it expires.
1. As an alternate path, can you get the DC's public certificate and use the `-Certificate` parameter?  It would need to be in CER Base64 format.

**Comment by erikgraa** [comment]
> There is no special negotiation here. This is dependent upon Microsoft's `System.Net.Sockets.TcpClient` class to download the certificate to then pass as a property value for the REST API.
> 
> Two things:
> 
> 1. Is there a reason why you are not adding your CA root chain as appliance trusted certificates?  By adding the Domain Controllers leaf cert, you then have to update the certificate from the DC when it expires.
> 2. As an alternate path, can you get the DC's public certificate and use the `-Certificate` parameter?  It would need to be in CER Base64 format.

1. is a good point. I can foresee issues when certificates invariably expire and one has forgotten the password of the local administrator. Do I not need the -TrustLeafCertificate in that case? It would be nice to figure out why this is failing nevertheless though, seeing as there is no issue ticking this checkbox off in the UI.

**Comment by ChrisLynchHPE** [comment]
No.  The `-TrustLeafCertificate` is only when you want to explicitly trust the certificate by adding it to the appliances trust store.  Otherwise, you are much better off here adding the CA root chain public Base64 certs to the appliance.  Then you do not need to explicitly trust the DC cert.  They will be implicitly trusted as long as they are valid and not revoked (which would need to be published via a CRL within the certs extended properties).

There isn't a way to instruct the appliance via the REST API to obtain the end points public certificate.  You need to provide it in the API call.  So the library needed a way to obtain the certificate, then convert to Base64 formatted data.

**Comment by erikgraa** [comment]
Thank you.

I tried using the Add-OVApplianceTrustedCertificate with the ComputerName and Port parameters, and the GET request seems to work fine (I see the certificate chain in the Request Body, i.e. CertificateInfoV2/CertificateDetails, but then I get this as it is POST-ing (seen with Verbose passed):

400 Bad Request
INVALID_JSON_MAPPING
JSON cannot be parsed
There is a problem on line 1 near column 5,293

Tried to put the JSON output that I saw when passing the Verbose flag into a PowerShell variable and piping it to ConvertFrom-Json, but got "unrecognised escape sequence". Will try to find out where it fails and add more info.

**Comment by ChrisLynchHPE** [comment]
In order to add the cert chain, you need to get that from the CA's themselves, or export the issuing CA from viewing the domain controllers public certificate, or using `openssl s_client -connect domaincontroller.domain.coml:636` and look for the base64 string value of its certificate.

If you happen to use openssl to get and then view the certificate within Windows, you can follow this screenshot to then export the issuing CA cert:

![Screenshot 2020-11-12 095530](https://user-images.githubusercontent.com/6956305/98977303-52bdd400-24cd-11eb-96aa-1cc887591170.png)

Once you copy the cert to a file (Base64 encoded), you can then use the [-Path](https://hpe-docs.gitbook.io/posh-hpeoneview/cmdlets/v5.50/appliance/add-ovappliancetrustedcertificate#path-less-than-system-io-fileinfo-greater-than) parameter with [Add-OVApplianceTrustedCertificate](https://hpe-docs.gitbook.io/posh-hpeoneview/cmdlets/v5.50/appliance/add-ovappliancetrustedcertificate).

**Comment by ChrisLynchHPE** [comment]
You don't use that Cmdlet to add the cert chain of the domain controller.  That functionality isn't there.  You need to get the domain controller cert chain and export/save the Base64 encoded value, then use the [-Path parameter](https://hpe-docs.gitbook.io/posh-hpeoneview/cmdlets/v5.50/appliance/add-ovappliancetrustedcertificate#path-less-than-system-io-fileinfo-greater-than) to [Add-OVApplianceTrustedCertificate](https://hpe-docs.gitbook.io/posh-hpeoneview/cmdlets/v5.50/appliance/add-ovappliancetrustedcertificate).

For instance, here is a screenshot of how to get that from Windows Explorer:
![Screenshot 2020-11-12 095530](https://user-images.githubusercontent.com/6956305/98982831-d3340300-24d4-11eb-9364-3e844998e52c.png)

**Comment by erikgraa** [comment]
Awesome, thanks for answering in detail!

I was able to use the cmdlet in the OP from a lower version Windows Server, so there's obviously some cipher setting that is mucking up with regard to no common algorithm being found. This is obviously not a bug with the library and as such I'll close this erroneous issue. 👍

**Comment by ChrisLynchHPE** [comment]
Thanks for reporting back.  There is an enhancement can be made here.  One that will help get the cert chain from an externally signed/issued cert authority. So I'm going to re-open this, and use it to track the enhancement.

**Comment by ChrisLynchHPE** [comment]
BTW, I'm going to add a new feature that will get the certificate chain from a provided certificate or SSL endpoint when not self-signed.  And will then add them to the appliance, instead of the leaf certificate.  But only when the caller provides a switch parameter, say `-AddCertChain`.

**Comment by erikgraa** [comment]
Splendid addition!

**Comment by erikgraa** [comment]
I tried the new ```Add-OVApplianceTrustedCertificate``` parameter ```-TrustCertChain``` along with ```-ComputerName <DC> -Port 636```and I am getting this error on usage:

![image](https://user-images.githubusercontent.com/20067955/99700401-6fd94080-2a93-11eb-909f-d52709dc3471.png)

**Comment by ChrisLynchHPE** [comment]
Can you run the command again, but within:

```powershell
Get-OVCommandTrace -ScriptBlock { 
    #Insert your code here
}
```

Then attach the generated file.

**Comment by ChrisLynchHPE** [comment]
Nevermind.  I found the source of the issue.  I had a merge problem from another library version, and caused some variables to get moved around.  I'll publish a quick fix for this, and some of the other items as soon as I get the 5.30 publishing going.  Which will be in the next hour or so.

**Comment by erikgraa** [comment]
I have tested the fix in 5.50.2607.2724 with great success.

**Comment by ChrisLynchHPE** [comment]
This is now addressed in the following releases:

* [5.50.2680.1687](https://www.powershellgallery.com/packages/HPEOneView.550/5.50.2680.1687) 
* [5.40.2681.1725](https://www.powershellgallery.com/packages/HPEOneView.540/5.40.2681.1725)
* [5.30.2682.3811](https://www.powershellgallery.com/packages/HPEOneView.530/5.30.2682.3811)
* [5.20.2684.2161](https://www.powershellgallery.com/packages/HPOneView.520/5.20.2684.2161)
* [5.00.2685.1865](https://www.powershellgallery.com/packages/HPOneView.500/5.0.2685.1865)

Closing.
